storage/cache/subtree_cache.go

```diff
@@ -169,36 +166,18 @@ func (s *SubtreeCache) cacheSubtree(t *storagepb.SubtreeProto) error {
 // GetNodes returns the requested nodes, calling the getSubtrees function if
 // they are not already cached.
 func (s *SubtreeCache) GetNodes(ids []compact.NodeID, getSubtrees GetSubtreesFunc) ([]tree.Node, error) {
-	if err := s.preload(ids, getSubtrees); err != nil {
+	if notFound, err := s.preload(ids, getSubtrees); err != nil {
 		return nil, err
+	} else if r := len(notFound); r != 0 {
+		return nil, fmt.Errorf("preload did not get all tiles: %d not found", r)
 	}
 
 	ret := make([]tree.Node, 0, len(ids))
 	for _, id := range ids {
-		h, err := s.getNodeHash(
-			id,
-			func(id []byte) (*storagepb.SubtreeProto, error) {
-				// This should never happen - we should've already read all the data we
-				// need above, in Preload()
-				glog.Warningf("Unexpectedly reading from within getNodeHash(): %x", id)
```

Comment:
Found our previous discussion on this: https://github.com/google/trillian/pull/2404#discussion_r596731648. Like before, I tried hard to understand whether/why this situation can occur, and concluded that it can't.

Every `id` from the request gets preloaded a few lines above (and if any tile is not found, an error is returned). So, if we reach this line here, we are sure that every relevant tile has been fetched and cached. If not, the new version of `getNodeHash` will return an error.

Reply:
It's possible that it only occurred in the map case. Just for clarity can you annotate the error that gets returned with something we can easily see in logs.

Reply:
Done.